### ISSUE TITLE
MRV2 `StagedWriteTensor.apply_write` computes row offsets in int32 and walks off the end of the buffer once `row_idx * stride(0) > 2**31`

### ISSUE THREAD

**Issue by Tsinghua-dhy** [vllm-project/vllm#57030]
We run a Qwen4Exp/QSA hybrid model at `max_model_len=1_000_000` (TP8, 8x MI308X, ROCm 7.2, torch 2.10.0+rocm7.2, main @ 9d14fd3095eb). Since mid-September the model gets dispatched to Model Runner V2, and every single engine died on the first real request with all TP workers taking a `Memory access fault by GPU node-N (Reason: Unknown)` within milliseconds of each other. We burned two days blaming CUDA graphs and sparse-attention kernels before finding the actual culprit here:

```python
# vllm/v1/worker/gpu/buffer_utils.py, _apply_write_kernel
row_ptr += row_idx * row_stride + start_idx
```

`row_idx` comes out of `write_indices` which is an int32 tensor, and `row_stride` is a host scalar (`self.gpu.stride(0)`) that Triton specializes to int32. So the multiply happens in int32. Any `StagedWriteTensor` where rows x stride crosses 2**31 gets a wrapped (negative) offset and the kernel stores out of bounds. On our config the product is ~2.6e9 — just past the line. Which is also why everything under 512k context is fine and 1M dies every time; it looks exactly like a "long context kernel bug" and nothing like an integer overflow.

Repro on any box with ~10GB free, no model needed:

```python
import torch
from vllm.v1.worker.gpu.buffer_utils import StagedWriteTensor

ROWS, COLS = 65536, 40000   # rows*stride = 2.62e9 > 2**31
swt = StagedWriteTensor((ROWS, COLS), torch.int32, torch.device("cuda"))
swt.stage_write(ROWS - 1, 0, [1, 2, 3, 4])
swt.apply_write()
torch.cuda.synchronize()
# expect gpu[ROWS-1, :4] == [1,2,3,4]; instead: illegal memory access / silent corruption
```

Worth noting the offset math can also *not* fault and just scribble somewhere else in the allocation — the silent-corruption variant is the one that should keep people up at night.

The fix we're running since yesterday across 6 production replicas (700k-token requests, FULL decode graphs, rock solid):

```diff
-    row_ptr += row_idx * row_stride + start_idx
+    row_ptr += row_idx.to(tl.int64) * row_stride + start_idx.to(tl.int64)
```

`_zero_kv_blocks_kernel` in `vllm/v1/worker/utils.py` already does the `.to(tl.int64)` dance, so the pattern is established; it just didn't reach this kernel. One thing to double check while in there: in the MULTI_GROUP path `row_stride` is loaded per-write from the `output_strides` tensor — worth confirming that tensor is int64 (or casting it too), otherwise the fused writer hits the same wall with slightly different numbers.

Can send a PR with the one-liner plus a size-guarded test if useful.

**Comment by vllmellm** [comment]
Hi @Tsinghua-dhy thank you for reporting this. We're able to reproduce the same errors on mi355x too. And we believed this is not limited to only AMD, the same int32 offset math applied to other accelerators' path too. 

The overflow is a combination of max-model-len × max-num-seqs exceeding 2^31 (MRV2's `free_indices.pop()` assigns the first request the highest row, `max_num_reqs - 1`). Could you share the `--max-num-seqs` you used? Concretely, in our case at 1M max-model-len, max-num-seqs > 2048 triggers the same fault but at server startup, since MRV2's kernel warmup drives the same `apply_write` path with synthetic requests.

Also, we found that the same overflow can be triggered by the in `_bincount_kernel` from `vllm/v1/worker/gpu/sample/penalties.py`, where it reads `req_state_idx * all_token_ids_stride`, and `PenaltiesState.apply_staged_writes` builds the mapping with an explicit dtype=torch.int32 — so with only the `buffer_utils.py` fix, the fault moves there. 

Perhaps you can open a PR on this?

**Comment by JiangLLM** [comment]
Thanks for the clear repro. I opened #57524 with the `int64` cast you suggested. It also fixes the same overflow in `_bincount_kernel` that @vllmellm pointed out.
